### ISSUE TITLE
Better child component reconciliation for web components.

### ISSUE THREAD

**Issue by wwwillchen** [google/mesop#449]
Right now, we do a naive approach were all the children elements are removed and then new children elements are attached. The problem is that this blows away the DOM state and is particularly annoying for components like checkbox that idiomatically rely on the DOM state for whether it's checked. You can workaround this by explicitly setting the `value`, but this isn't ideal.

We can also improve the performance by having a more sophisticated way of reconciling child components.

**Comment by richard-to** [comment]
Yeah, I just ran into this issue today. I was experimenting with a hot keys web component that could wrap say a core textarea or even the whole Mesop app page. But yeah, since I needed to use on_input to capture the text, every couple key presses would reset the text area and make me lose focus of the text area.

Here's the example branch: https://github.com/google/mesop/compare/main...richard-to:mesop:async-action-hot-keys

That commit also includes another experimental component (async action component) which can be ignored. It's mainly the hot key one that is of interest in this scenario.
